### PR TITLE
Add Okta as authentication backend.

### PR DESCRIPTION
**Pull request by glogiotatidis**
- [x] update remote_metadata.xml before merging

### CODE REVIEW

**Comment by glogiotatidis** [comment]
This implements Okta on careers. I've a demo instance on https://careers-giorgos.us-west.moz.works which authenticates against oktapreview. Ping me on IRC for more info.

@jgmize r?

**Comment by jgmize** [comment]
I tested this by trying to sign in to the admin on the demo instance, went through the sign-in process (including selecting a "security image" and MFA) but at the end of what appeared to be a successful login I ended up on https://mozilla.oktapreview.com/app/UserHome# instead of https://careers-giorgos.us-west.moz.works/admin/ and when I manually visited https://careers-giorgos.us-west.moz.works/admin/ it still prompted me to login. Given that my account didn't actually have admin on the instance at that point, this seemed reasonable even if the UX could be improved. I then connected to a shell on the demo instance via 'deisctl dock' and created a superuser with my email address and tried logging in again, but now I'm getting a message from mozilla.oktapreview.com: "Sorry, you can't access Careers Demo because you are not assigned this app in Okta."

File: careers/saml/remote_metadata.xml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ZxRBhH7WwhWY4jIOETSYZY62d2amUlHegvqMVuXpc3wjdxD3YV+2nILad6HLppCKR53cq+VDh6rO
+AxDSSHYGfqXIWFDGOTOpSdoNnmy6swIDAQABMA0GCSqGSIb3DQEBBQUAA4IBAQAkvhAmawj5ptqu
+9fg74npE64shfHguW4GKmG80Zc9fshUX057eIfNsNOEl+cnULJ9X02umld6VEbv3BTFLIpE/iK3i
+rwmTVLp27u2FxHo0qm7bs1sRZIsSQ7dvnIX3M++tcKqKdEIDjNvAM5IgB0xph2KQ6bkqovWOceHv
+9G2eDrGqEfmL44pASZ8xwY3kn6qUUY1hSjWI3pkROTeW+VCCR/gA1/I0PhaNcT5UoZ7b2F0ptGsu
+CGq4KkDUXXTIHRl+f3VNDPJBjxVZBpkpYEdNFnhcmWADwNgtmYwW4rYgb4s7VaGA5Hg75Z3NPlR4
+Brq4s05ixFJAciflA4b7mnic</ds:X509Certificate></ds:X509Data></ds:KeyInfo></md:KeyDescriptor><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat><md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://mozilla.oktapreview.com/app/mozilla_careersdemo_1/exk5q0r144JuD8UJr0h7/sso/saml"/><md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://mozilla.oktapreview.com/app/mozilla_careersdemo_1/exk5q0r144JuD8UJr0h7/sso/saml"/></md:IDPSSODescriptor></md:EntityDescriptor>
```
Comment (jgmize) [inline careers/saml/remote_metadata.xml]: How was this certificate generated?

Comment (glogiotatidis) [inline careers/saml/remote_metadata.xml]: That's coming from Okta and it's the cert it uses to sign it's asserttions.

**Comment by glogiotatidis** [comment]
> I tested this by trying to sign in to the admin on the demo instance, went through the sign-in process (including selecting a "security image" and MFA) but at the end of what appeared to be a successful login I ended up on https://mozilla.oktapreview.com/app/UserHome# instead of https://careers-giorgos.us-west.moz.works/admin/ and when I manually visited https://careers-giorgos.us-west.moz.works/admin/ it still prompted me to login. Given that my account didn't actually have admin on the instance at that point, this seemed reasonable even if the UX could be improved. I then connected to a shell on the demo instance via 'deisctl dock' and created a superuser with my email address and tried logging in again, but now I'm getting a message from mozilla.oktapreview.com: "Sorry, you can't access Careers Demo because you are not assigned this app in Okta."

Hey I told you to ping me on IRC to avoid all this trouble :) I assigned the app to you, to you should be set. For demo purposes this automatically creates accounts, which don't have full permissions and therefore cannot login on /admin as you correctly noted, but in production you'd get a "Access denied" message unless there's an account for you in the django db.

**Comment by jgmize** [comment]
Nice work, @glogiotatidis. As we discussed, I think the next steps are to add the `auth.backends.ModelBackend` based on an environment variable and to generate and update the remote_metadata.xml from production okta. We should have someone from the security group who is more familiar with saml take a look at this and make sure we haven't overlooked anything.

**Comment by gene1wood** [comment]
I believe that this *adds* Okta as an authentication method to Django, but doesn't disable the [built in authentication endpoints](https://github.com/glogiotatidis/lumbergh/blob/okta/careers/urls.py#L11). Is this intentional? I think that an attacker could still interact with these endpoints and attempt to brute force a user in the DB that still has a password (e.g. an older user, pre-cutover-to-okta).

Other than that, looks good to me.

**Comment by glogiotatidis** [comment]
> I believe that this adds Okta as an authentication method to Django, but doesn't disable the built in authentication endpoints. Is this intentional? I think that an attacker could still interact with these endpoints and attempt to brute force a user in the DB that still has a password (e.g. an older user, pre-cutover-to-okta).

We still need access to the admin panel, that's the only place that login is required anyway. But we do not support authenticating with username and password anymore, since we limit [Django's Authentication Backends to Okta](https://github.com/mozilla/lumbergh/pull/154/files#diff-984b850e779ad4e8b6a06be805f70a55R15) and we [remove the username / passwd form from the admin login page](https://github.com/mozilla/lumbergh/pull/154/files#diff-d942f02b2ed2a30cc0226a33f6cb8ce3R32) when SAML is enabled.  Therefore attackers cannot brute force username's and password's.

Thank you for your time @gene1wood

**Comment by glogiotatidis** [comment]
Addressed all comments, tests pass, merging this. :ship:

**Comment by glogiotatidis** [comment]
@jgmize this is now on production. Can you please verify that you're able to login on https://careers.mozilla.org/admin ?

**Comment by jgmize** [comment]
> @jgmize this is now on production. Can you please verify that you're able to login on https://careers.mozilla.org/admin?

Verified. Nice work, @glogiotatidis!

**Comment by gene1wood** [comment]
> we do not support authenticating with username and password anymore, since we limit Django's Authentication Backends to Okta

Aha, got it. Yes, that looks perfect. I hadn't noticed the absent `django.contrib.auth.backends.ModelBackend` from the `AUTHENTICATION_BACKENDS` list.

**Pull request merged**